moveit_ros/trajectory_cache/src/trajectory_cache.cpp

```diff
@@ -0,0 +1,1340 @@
+
+      // Compute offsets wrt. to workspace frame.
+      geometry_msgs::msg::Quaternion quat_offset;
+      quat_offset.x = 0;
+      quat_offset.y = 0;
+      quat_offset.z = 0;
+      quat_offset.w = 1;
+
+      if (plan_request.workspace_parameters.header.frame_id != constraint.header.frame_id)
+      {
+        try
+        {
+          auto transform = tf_buffer_->lookupTransform(
+              constraint.header.frame_id, plan_request.workspace_parameters.header.frame_id, tf2::TimePointZero);
+
+          quat_offset = transform.transform.rotation;
+        }
+        catch (tf2::TransformException& ex)
+        {
+          RCLCPP_WARN(node_->get_logger(),
+                      "Skipping goal metadata append: "
+                      "Could not get goal transform for orientation %s to %s: %s",
+                      plan_request.workspace_parameters.header.frame_id.c_str(), constraint.header.frame_id.c_str(),
+                      ex.what());
+
+          // (Can't. Copy not supported.)
+          // *metadata = original;  // Undo our changes.
+          return false;
+        }
+      }
+
+      metadata.append(meta_name + ".link_name", constraint.link_name);
+
+      // Orientation of constraint frame wrt. workspace frame
+      tf2::Quaternion tf2_quat_frame_offset(quat_offset.x, quat_offset.y, quat_offset.z, quat_offset.w);
+
+      // Added offset on top of the constraint frame's orientation stated in
+      // the constraint.
+      tf2::Quaternion tf2_quat_goal_offset(constraint.orientation.x, constraint.orientation.y, constraint.orientation.z,
+                                           constraint.orientation.w);
+
+      tf2_quat_frame_offset.normalize();
+      tf2_quat_goal_offset.normalize();
+
+      auto final_quat = tf2_quat_goal_offset * tf2_quat_frame_offset;
+      final_quat.normalize();
+
+      metadata.append(meta_name + ".target_point_offset.x", final_quat.getX());
+      metadata.append(meta_name + ".target_point_offset.y", final_quat.getY());
+      metadata.append(meta_name + ".target_point_offset.z", final_quat.getZ());
+      metadata.append(meta_name + ".target_point_offset.w", final_quat.getW());
+    }
+  }
+
+  return true;
+}
+
+// =============================================================================
+// CARTESIAN TRAJECTORY CACHING
+// =============================================================================
+// CARTESIAN TRAJECTORY CACHING: TOP LEVEL OPS
+moveit_msgs::srv::GetCartesianPath::Request
+TrajectoryCache::construct_get_cartesian_path_request(moveit::planning_interface::MoveGroupInterface& move_group,
+                                                      const std::vector<geometry_msgs::msg::Pose>& waypoints,
+                                                      double step, double jump_threshold, bool avoid_collisions)
+{
+  moveit_msgs::srv::GetCartesianPath::Request out;
+
+  // Some of these parameters need us to pull PRIVATE values out of the
```

Comment:
Feel free to address this in this PR and make changes to the move_group too :+1:

Reply:
Added in: https://github.com/moveit/moveit2/pull/2925

I will rebase this PR's branch on top of that PR's branch.

Reply:
Used in f6151e38b08e51e30584946f56ae189a9dacea1f, but depends on #2925